File: packages/sdkwork-utils-python/sdkwork_utils/encoding.py
import base64
import binascii
from urllib.parse import quote, unquote


def base64_encode(value: bytes | str) -> str:
    data = value.encode("utf-8") if isinstance(value, str) else value
    return base64.b64encode(data).decode("ascii")


def base64_decode(value: str) -> bytes | None:
    try:
        return base64.b64decode(value.strip(), validate=True)
    except (binascii.Error, ValueError):
        return None


def hex_encode(value: bytes) -> str:
    return value.hex()


def hex_decode(value: str) -> bytes | None:
    trimmed = value.strip()
    if len(trimmed) % 2 != 0:
        return None
    try:
        return bytes.fromhex(trimmed)
    except ValueError:
        return None


def url_encode(value: str) -> str:
    return quote(value, safe="")


def url_decode(value: str) -> str | None:
    try:
        return unquote(value)
    except ValueError:
        return None


def base64url_encode(value: bytes | str) -> str:
    data = value.encode("utf-8") if isinstance(value, str) else value
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def base64url_decode(value: str) -> bytes | None:
    trimmed = value.strip()
    if not trimmed:
        return None
    padding = "=" * ((4 - len(trimmed) % 4) % 4)
    try:
        return base64.urlsafe_b64decode(trimmed + padding)
    except (binascii.Error, ValueError):
        return None

File: packages/sdkwork-utils-typescript/src/encoding.ts
import {
  base64Decode as decodeBase64,
  base64Encode as encodeBase64,
  base64UrlDecode as decodeBase64Url,
  base64UrlEncode as encodeBase64Url,
  hexDecode as decodeHex,
  hexEncode as encodeHex,
  toUtf8,
} from "./runtime/binary.js";

export function base64Encode(value: string | Uint8Array): string {
  const bytes = typeof value === "string" ? toUtf8(value) : value;
  return encodeBase64(bytes);
}

export function base64Decode(value: string): Uint8Array | null {
  return decodeBase64(value);
}

export function hexEncode(value: Uint8Array): string {
  return encodeHex(value);
}

export function hexDecode(value: string): Uint8Array | null {
  return decodeHex(value);
}

export function urlEncode(value: string): string {
  return encodeURIComponent(value);
}

export function urlDecode(value: string): string | null {
  try {
    return decodeURIComponent(value);
  } catch {
    return null;
  }
}

export function base64UrlEncode(value: string | Uint8Array): string {
  const bytes = typeof value === "string" ? toUtf8(value) : value;
  return encodeBase64Url(bytes);
}

export function base64UrlDecode(value: string): Uint8Array | null {
  return decodeBase64Url(value);
}
